### PR TITLE
Webcmdlets display an error on https to http redirect

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary

At the moment https to http redirect fail silently, after this PR it will show an error.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx [hunk 1]
```diff
@@ -159,12 +159,15 @@
...
     <value>Cannot convert the JSON string because a dictionary that was converted from the string contains the duplicated key '{0}'.</value>
   </data>
-  <data name="KeysWithDifferentCasingInJsonString" xml:space="preserve">
-    <value>Cannot convert the JSON string because it contains keys with different casing. Please use the -AsHashTable switch instead. The key that was attempted to be added to the existing key '{0}' was '{1}'.</value>
-  </data>
   <data name="IEDomNotSupported" xml:space="preserve">
     <value>The response content cannot be parsed because the Internet Explorer engine is not available, or Internet Explorer's first-launch configuration is not complete. Specify the UseBasicParsing parameter and try again. </value>
   </data>
+  <data name="InsecureRedirection" xml:space="preserve">
+    <value>Cannot follow an insecure redirection by default. Reissue the command specifying the AllowInsecureRedirect parameter. </value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx:166]:
```suggestion
    <value>Cannot follow an insecure redirection by default. Reissue the command specifying the AllowInsecureRedirect switch. </value>
```

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx]: Done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 2]
```diff
@@ -1550,6 +1550,14 @@ protected override void ProcessRecord()
                                     OutFile = null;
                                 }
 
+                                // Detect insecure redirection
+                                if (response.RequestMessage.RequestUri.Scheme == "https" && response.Headers.Location?.Scheme == "http")
+                                {
+                                        ErrorRecord er = new(new InvalidOperationException(), "InsecureRedirection", ErrorCategory.InvalidOperation, request);
+                                        er.ErrorDetails = new ErrorDetails(WebCmdletStrings.InsecureRedirection);
+                                        WriteError(er);
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1558]: Shouldn't this be a terminating error here?  There doesn't seem to be any point in continuing and we should throw `ThrowTerminatingError()` here.

**Review by PaulHigin — DISMISSED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 2 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1558]: I used `WriteError()` to mirror the behaviour of lines 1604-1609:

```cs
if (WebSession.MaximumRedirection == 0 && IsRedirectCode(response.StatusCode)) // Indicate "HttpClientHandler.AllowAutoRedirect == false"
{
    ErrorRecord er = new(new InvalidOperationException(), "MaximumRedirectExceeded", ErrorCategory.InvalidOperation, request);
    er.ErrorDetails = new ErrorDetails(WebCmdletStrings.MaximumRedirectionCountExceeded);
    WriteError(er);
}
```
I can change it if needed.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1558]: What is a behavior of other utils in the scenario?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1558]: Also text in WebCmdletStrings.InsecureRedirection assumes we should throw.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1558]: Test:
`Invoke-WebRequest "https://mockbin.org/redirect/308?to=https://mockbin.org/redirect/302?to=http://mockbin.org/status/200" -SkipHttpErrorCheck` --> 302
`curl -L "https://mockbin.org/redirect/308?to=https://mockbin.org/redirect/302?to=http://mockbin.org/status/200"` --> 200

(HttpClient does not seem to allow insecure redirect: https://github.com/dotnet/runtime/issues/23801)

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Please add new test.

**Comment by iSazonov** [comment]
@CarloToso Please add tests.

**Comment by CarloToso** [comment]
@iSazonov I'm working on the tests

**Comment by iSazonov** [comment]
@CarloToso We could add tests in #18546 after current PR will be merged.

**Comment by CarloToso** [comment]
@iSazonov If the tests pass I think we can merge

**Review by iSazonov — APPROVED** [review]

**Comment by CarloToso** [comment]
@SteveL-MSFT can we merge this?

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by crlogic** [comment]
> **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
> * [x]  None

Hrmm, this breaks every IRM across corporate zero-trust proxies by making `AllowInsecureRedirect` mandatory (because I cannot control this as an end user).  Except I can't perform a blanket "add AllowInsecureRedirect" to all IRM's because agents/executors/etc won't all be upgraded big-bang to 7.4 (rolling upgrades). 

Script modules are now completely broken if there is a insecure proxy.  Wouldn't that make this a breaking change?
https://github.com/dell/OpenManage-PowerShell-Modules

Is there an easier way?
```
    if ($psVersion -ge 7.4.0) {
        $querySplat.Add('AllowInsecureRedirect',$true)
    }
```

**Comment by CarloToso** [comment]
@crlogic Could you please open an issue to discuss this?

**Comment by crlogic** [comment]
> https://github.com/dell/OpenManage-PowerShell-Modules

I hope I did it right, https://github.com/PowerShell/PowerShell/issues/19461

**Pull request merged**